### PR TITLE
chore: convert policy.js to ts

### PR DESCRIPTION
**Pull request by lwywoo**
- [x] Ready for review
- [x] Follows [CONTRIBUTING](https://github.com/snyk/snyk/blob/master/.github/CONTRIBUTING.md) rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?
Convert policy.js and policy test to ts.

### CODE REVIEW

File: src/lib/display-policy.ts [hunk 1]
```diff
@@ -1,10 +1,8 @@
...
+import {demunge} from 'snyk-policy';
+import * as config from './config';
 
-const chalk = require('chalk');
-const demunge = require('snyk-policy').demunge;
-const config = require('./config');
-
-function display(policy) {
+export async function display(policy) {
   return new Promise(((resolve) => {
```
Comment (dkontorovskyy) [inline src/lib/display-policy.ts]: This Promise could be removed. Just execute code and in the end to `return Promise.resolve(res)` or `await res`. Should work, but check :)

File: test/policy.test.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
+import * as policy from 'snyk-policy';
+import * as test from 'tape';
+import {extendExpiries} from './utils';
+import * as path from 'path';
+
+test('policy is callable', (t) => {
+  const dir = path.resolve(__dirname + '/fixtures/jsbin-snyk-config');
+  t.plan(3);
```
Comment (dkontorovskyy) [inline test/policy.test.ts]: As much as I remember, you can remove all `t.plan(n)`, just add `async` in line 6 before `(t)`

File: test/policy.test.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  policy.load(dir).then(() => {
```
Comment (dkontorovskyy) [inline test/policy.test.ts]: do it `async/await`

File: test/policy.test.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    t.pass('called with string');
+  });
+
+  policy.load([dir]).then(() => {
```
Comment (dkontorovskyy) [inline test/policy.test.ts]: do it `async/await`

File: test/policy.test.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    t.pass('called with array');
+  });
+
+  policy.load(dir, { 'ignore-policy': true }).then(() => {
```
Comment (dkontorovskyy) [inline test/policy.test.ts]: do it `async/await`

File: test/policy.test.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+test('test sensibly bails if gets an old .snyk format', (t) => {
+  const dir = path.resolve(__dirname + '/fixtures/jsbin-snyk-config');
+  const vulns = require('./fixtures/test-jsbin-vulns-updated.json');
+  const id = 'npm:semver:20150403';
+  const vuln = vulns.vulnerabilities.filter((v) => {
+    return v.id === id;
+  }).pop();
+
+  policy.load(dir).then((config) => {
```
Comment (dkontorovskyy) [inline test/policy.test.ts]: do it `async/await`

File: test/policy.test.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    t.fail('could not load the policy file');
+    t.end();
+  });
+});
+
+test('policy ignores correctly', (t) => {
+  const dir = path.resolve(__dirname + '/fixtures/hapi-azure-post-update/');
+  const vulns = require(dir + '/test.json');
+
+  policy.load(dir).then((config) => {
```
Comment (dkontorovskyy) [inline test/policy.test.ts]: do it `async/await`

File: test/policy-display.test.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+test('test sensibly bails if gets an old .snyk format', async (t) => {
+
+  t.plan(1);
+  const filename = __dirname + '/fixtures/snyk-config-no-version';
+  const promises = [
+    policy.load(filename).then(display),
+    fs.readFile(filename + '/expected', 'utf8'),
+  ];
+  try {
+     await Promise.all(promises).then((res) => {
```
Comment (dkontorovskyy) [inline test/policy-display.test.ts]: `await Promise.all` will return you array of resolved Promises. So you can refactor this to `async/await`

File: test/policy-display.test.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
+import * as policy from 'snyk-policy';
+import {test} from 'tap';
+import * as fs from 'then-fs';
+import {display} from '../src/lib/display-policy';
+import stripAnsi from 'strip-ansi';
+
+test('test sensibly bails if gets an old .snyk format', async (t) => {
+
+  t.plan(1);
```
Comment (dkontorovskyy) [inline test/policy-display.test.ts]: if you have `async` function this is not needed, but check it up please :)

File: src/lib/display-policy.ts (hunk 1 above)
Comment (kyegupov) [reply src/lib/display-policy.ts]: yeah, the outer async should take care of this.

File: src/lib/display-policy.ts [hunk 10]
```diff
@@ -24,14 +22,14 @@ function display(policy) {
 }
 
 function displayRule(title) {
-  return function (rule, i) {
+  return (rule, i) => {
```
Comment (kyegupov) [inline src/lib/display-policy.ts]: I think you can pull this helper function outside of the outer function, it does not appear to be a closure.

File: test/policy-display.test.ts (hunk 9 above)
Comment (kyegupov) [reply test/policy-display.test.ts]: +1

File: test/policy.test.ts [hunk 11]
```diff
@@ -0,0 +1,83 @@
+import * as policy from 'snyk-policy';
+import * as test from 'tape';
+import {extendExpiries} from './utils';
+import * as path from 'path';
+
+test('policy is callable with string', async (t) => {
+  const dir = path.resolve(__dirname + '/fixtures/jsbin-snyk-config');
+  try {
+    await policy.load(dir).then(() => {
```
Comment (dkontorovskyy) [inline test/policy.test.ts]: you don't need then here.
```
await policy.load(dir);
t.pass('called with string');
t.end();
```

File: test/policy.test.ts [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    });
+  } catch (e) {
+    t.fail(e);
+  }
+});
+
+test('policy is callable with array', async (t) => {
+  const dir = path.resolve(__dirname + '/fixtures/jsbin-snyk-config');
+  try {
+    await policy.load([dir]).then(() => {
```
Comment (dkontorovskyy) [inline test/policy.test.ts]: no need in then. 

```
await policy.load([dir]);
t.pass('called with array');
t.end();
```

File: test/policy.test.ts [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+    });
+  } catch (e) {
+    t.fail(e);
+  }
+});
+
+test('policy is callable with string and options', async (t) => {
+  const dir = path.resolve(__dirname + '/fixtures/jsbin-snyk-config');
+  try {
+    await policy.load(dir, { 'ignore-policy': true }).then(() => {
```
Comment (dkontorovskyy) [inline test/policy.test.ts]: same as above

File: test/policy.test.ts [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+
+test('test sensibly bails if gets an old .snyk format', async (t) => {
+  const dir = path.resolve(__dirname + '/fixtures/jsbin-snyk-config');
+  const vulns = require('./fixtures/test-jsbin-vulns-updated.json');
+  const id = 'npm:semver:20150403';
+  const vuln = vulns.vulnerabilities.filter((v) => {
+    return v.id === id;
+  }).pop();
+  try {
+    await policy.load(dir).then((config) => {
```
Comment (dkontorovskyy) [inline test/policy.test.ts]: same as above

File: test/policy.test.ts [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+      t.end();
+  }
+});
+
+test('policy ignores correctly', async (t) => {
+  const dir = path.resolve(__dirname + '/fixtures/hapi-azure-post-update/');
+  const vulns = require(dir + '/test.json');
+
+  try {
+    await policy.load(dir).then((config) => {
```
Comment (dkontorovskyy) [inline test/policy.test.ts]: same as above

File: test/policy-display.test.ts [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+import stripAnsi from 'strip-ansi';
+
+test('test sensibly bails if gets an old .snyk format', async (t) => {
+
+  const filename = __dirname + '/fixtures/snyk-config-no-version';
+  const loadedPolicy = await policy.load(filename);
+  const expectedFile = await fs.readFile(filename + '/expected', 'utf8');
+
+  try {
+     return await Promise.all([display(loadedPolicy), expectedFile]).then((res) => {
```
Comment (dkontorovskyy) [inline test/policy-display.test.ts]: same goes here.
```
const [displayPolicy, expectedFileString] = await Promise.all([display(loadedPolicy), expectedFile]);
const result = stripAnsi(displayPolicy).trim().split('\n').slice(3).join('\n');
const expected = expectedFileString.trim().split('\n').slice(3).join('\n');
t.equal(result, expected);
```

File: src/lib/display-policy.ts (hunk 10 above)
Comment (lwywoo) [reply src/lib/display-policy.ts]: It looks like it is a closure on `title`, would you be able to explain what you mean?

Comment (kyegupov) [reply src/lib/display-policy.ts]: Yes, you're right.

File: src/cli/commands/policy.ts [hunk 17]
```diff
@@ -0,0 +1,18 @@
+import * as policy from 'snyk-policy';
+import {display} from '../../lib/display-policy';
+import * as errors from '../../lib/errors';
```
Comment (dkontorovskyy) [inline src/cli/commands/policy.ts]: It's more a suggestion or open question
Would it be nicer to have `import {FailedToLoadPolicyError, PolicyNotFoundError} from '../../lib/errors'`?

**Pull request merged**